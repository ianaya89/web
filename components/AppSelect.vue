<template>
  <div class="app_select">
    <span v-if="soloLabel" class="text-grey text-p3 uppercase tracking-wide">
      {{ soloLabel }}
    </span>
    <span v-if="soloLabel && required" class="text-red text-p3">
      *
    </span>
    <app-tooltip v-if="soloLabel && tooltip" class="text-p2">
      {{ tooltipText }}
    </app-tooltip>
    <v-select
      ref="select"
      v-bind="$props"
      :outline="outline"
      :flat="true"
      :append-icon="'fa-chevron-down'"
      :error-messages="translatableErrorMessages"
      @input="onSelect"
    >
      <template v-if="loading" v-slot:no-data>
        <div class="mt-5 box-content h-8 w-8 w-full text-center">
          {{ $t('general.loading_data') }}
        </div>
      </template>
    </v-select>
  </div>
</template>

<script>
import BaseInput from '@/components/BaseInput'
import AppTooltip from '@/components/AppTooltip'

export default BaseInput.extend({
  name: 'AppSelect',

  components: {
    AppTooltip
  },

  props: {
    soloLabel: {
      type: String,
      default: 'Label'
    },

    label: {
      type: String,
      default: null
    },

    items: {
      type: Array,
      default: () => []
    },
    outline: {
      type: Boolean,
      default: true
    },
    loading: {
      type: Boolean,
      default: false
    },
    value: {
      type: [String, Number, Object],
      default: null
    },
    tooltip: {
      type: Boolean,
      default: null
    },
    errorMessages: {
      type: Array,
      default() {
        return []
      }
    },
    tooltipText: {
      type: String,
      default: 'Tooltip'
    }
  },

  computed: {
    translatableErrorMessages() {
      return this.errorMessages.map(message => this.$t(message))
    },

    hasError() {
      if (!this.$refs.select) {
        return false
      }

      return this.$refs.select.hasError
    }
  },

  data: () => ({
    selected: []
  }),

  methods: {
    onSelect(value) {
      this.$emit('input', value)
    },

    resetValidation() {
      if (this.$refs.select) {
        this.$refs.select.resetValidation()
      }
    },

    validate() {
      if (this.$refs.select) {
        this.$refs.select.validate(true)
      }
    }
  }
})
</script>

<style lang="postcss">
.app_select {
  .v-icon {
    @apply .text-red .text-menu .font-icon;
  }

  .v-select__selection--comma {
    margin: 0.55rem 0.25rem 0.5rem 0;
  }

  .v-list {
    @apply .p-0;
  }

  .v-menu__content {
    top: 0.375rem !important;
    left: 0.125rem !important;
    @apply .rounded-none .border .border-grey-l2;
  }
  .v-input__slot {
    @apply .mb-0 .border .border-grey-l2 .rounded-none !important;

    &::before {
      @apply .hidden;
    }
  }

  .v-list__tile {
    border-bottom: 1px solid lightgrey;
    @apply .flex-row-reverse;
  }

  .v-list__tile__action {
    @apply .min-w-0;
  }

  .v-list__tile--active {
    @apply .font-bold;
  }
}
.v-select.v-text-field--enclosed:not(.v-text-field--single-line) {
  .v-select__selections {
    padding-top: 0px;
  }
}
</style>
