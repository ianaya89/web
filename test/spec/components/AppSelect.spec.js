import { mount, shallowMount } from '@vue/test-utils'
import AppSelect from '@/components/AppSelect'

describe('AppSelect', () => {
  it('renders default select', () => {
    const wrapper = shallowMount(AppSelect)
    expect(wrapper.element).toMatchSnapshot()
  })

  it('renders some props', () => {
    const wrapper = shallowMount(AppSelect, {
      propsData: {
        items: [
          {
            text: 'First',
            value: 0
          },
          {
            text: 'Second',
            value: 1
          }
        ],
        disabled: true
      }
    })
    expect(wrapper.element).toMatchSnapshot()
  })

  it('set value', () => {
    const wrapper = shallowMount(AppSelect, {
      propsData: {
        items: [
          {
            text: 'First',
            value: 0
          },
          {
            text: 'Second',
            value: 1
          }
        ]
      }
    })

    expect(wrapper.element).toMatchSnapshot()

    wrapper.setProps({
      value: 1
    })

    expect(wrapper.element).toMatchSnapshot()
  })

  it('emit input event', () => {
    const wrapper = mount(AppSelect, {
      propsData: {
        items: [
          {
            text: 'First',
            value: 0
          },
          {
            text: 'Second',
            value: 1
          }
        ]
      }
    })

    wrapper.vm.onSelect(1)

    expect(wrapper.emitted('input')).toBeTruthy()
    expect(wrapper.emitted('input')[0][0]).toEqual(1)
  })

  it('it validate input field', async () => {
    const wrapper = mount(AppSelect, {
      propsData: {
        required: true,
        rules: [v => !!v || 'required'],
        items: [
          {
            text: 'First',
            value: 0
          },
          {
            text: 'Second',
            value: 1
          }
        ]
      }
    })

    wrapper.vm.validate()

    expect(wrapper.vm.$refs.select.valid).toBeFalsy()

    await wrapper.setProps({
      value: 1
    })

    wrapper.vm.validate()

    expect(wrapper.vm.$refs.select.valid).toBeTruthy()
  })

  it('it reset validation', () => {
    const wrapper = mount(AppSelect, {
      propsData: {
        required: true,
        rules: [v => !!v || 'required'],
        items: [
          {
            text: 'First',
            value: 0
          },
          {
            text: 'Second',
            value: 1
          }
        ]
      }
    })

    wrapper.vm.validate()

    expect(wrapper.vm.$refs.select.hasMessages).toBeTruthy()

    wrapper.vm.resetValidation()

    expect(wrapper.vm.$refs.select.hasMessages).toBeFalsy()
  })
})
