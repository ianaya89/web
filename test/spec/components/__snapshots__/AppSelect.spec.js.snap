// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppSelect renders default select 1`] = `
<div
  class="app_select"
>
  <span
    class="text-grey text-p3 uppercase tracking-wide"
  >
    
    Label
  
  </span>
   
  <!---->
   
  <!---->
   
  <v-select-stub
    append-icon="fa-chevron-down"
    error-messages=""
    flat="true"
    items=""
    outline="true"
    rules=""
    sololabel="Label"
    tooltiptext="Tooltip"
  />
</div>
`;

exports[`AppSelect renders some props 1`] = `
<div
  class="app_select"
>
  <span
    class="text-grey text-p3 uppercase tracking-wide"
  >
    
    Label
  
  </span>
   
  <!---->
   
  <!---->
   
  <v-select-stub
    append-icon="fa-chevron-down"
    disabled="true"
    error-messages=""
    flat="true"
    items="[object Object],[object Object]"
    outline="true"
    rules=""
    sololabel="Label"
    tooltiptext="Tooltip"
  />
</div>
`;

exports[`AppSelect set value 1`] = `
<div
  class="app_select"
>
  <span
    class="text-grey text-p3 uppercase tracking-wide"
  >
    
    Label
  
  </span>
   
  <!---->
   
  <!---->
   
  <v-select-stub
    append-icon="fa-chevron-down"
    error-messages=""
    flat="true"
    items="[object Object],[object Object]"
    outline="true"
    rules=""
    sololabel="Label"
    tooltiptext="Tooltip"
  />
</div>
`;

exports[`AppSelect set value 2`] = `
<div
  class="app_select"
>
  <span
    class="text-grey text-p3 uppercase tracking-wide"
  >
    
    Label
  
  </span>
   
  <!---->
   
  <!---->
   
  <v-select-stub
    append-icon="fa-chevron-down"
    error-messages=""
    flat="true"
    items="[object Object],[object Object]"
    outline="true"
    rules=""
    sololabel="Label"
    tooltiptext="Tooltip"
  />
</div>
`;
